writer/writer.go

```diff
@@ -15,27 +15,380 @@
+	"data/*": {Path: "data/*", IsDir: false, Required: true},
+}
+
+// NewArtifactsWriter creates a new ArtifactsWriter providing a location
+// of Mender metadata artifacts, format of the update and version.
+func NewArtifactsWriter(path string, format string, version int) *ArtifactsWriter {
+	return &ArtifactsWriter{
+		updateLocation:  path,
+		headerStructure: metadata.MetadataArtifactHeader{Artifacts: ArtifactsHeaderFormat},
+		format:          format,
+		version:         version,
+		updates:         make(map[string]updateBucket),
+	}
+}
+
+type updateArtifact struct {
+	name         string
+	path         string
+	updateBucket string
+	info         os.FileInfo
+	checksum     []byte
+}
+
+type updateBucket struct {
+	location        string
+	path            string
+	archivedPath    string
+	updateArtifacts []updateArtifact
+	files           metadata.MetadataFiles
+}
+
+// ReadArchiver provides interface for reading files or streams and preparing
+// those to be written to tar archive.
+// GetHeader returns Header to be written to the crrent entry it the tar archive.
+type ReadArchiver interface {
+	io.ReadCloser
+	Open() error
+	GetHeader() (*tar.Header, error)
+}
+
+func (av ArtifactsWriter) calculateChecksum(upd *updateArtifact) error {
+	f, err := os.Open(upd.path)
+	if err != nil {
+		return errors.Wrapf(err, "can not open file for calculating checksum")
+	}
+	defer f.Close()
+	h := sha256.New()
+	if _, err := io.Copy(h, f); err != nil {
+		return errors.Wrapf(err, "error calculating checksum")
+	}
+
+	checksum := h.Sum(nil)
+	upd.checksum = make([]byte, hex.EncodedLen(len(checksum)))
+	hex.Encode(upd.checksum, h.Sum(nil))
+	log.Debugf("hash of file: %v (%x)\n", upd.path, upd.checksum)
+	return nil
+}
+
+func (av ArtifactsWriter) getHeaderInfo(updates []os.FileInfo) metadata.MetadataHeaderInfo {
+	// for now we have ONLY one type of update - rootfs-image
+	headerInfo := metadata.MetadataHeaderInfo{}
+
+	// TODO: should we store update name as well?
+	for _ = range updates {
+		headerInfo.Updates = append(headerInfo.Updates, metadata.MetadataUpdateType{Type: "rootfs-image"})
 	}
-	if err := info.Validate(); err == metadata.ErrInvalidInfo {
-		return nil, nil
+	return headerInfo
+}
+
+func (av ArtifactsWriter) getInfo() metadata.MetadataInfo {
+	return metadata.MetadataInfo{
+		Format:  av.format,
+		Version: av.version,
+	}
+}
+
+func (av *ArtifactsWriter) writeArchive(destination io.WriteCloser, content []ReadArchiver, compressed bool) error {
+	if len(content) == 0 {
+		return errors.New("artifacts writer: empty content")
+	}
+
+	var tw *tar.Writer
+	if compressed {
+		// start with something simple for now
+		gz := gzip.NewWriter(destination)
+		defer gz.Close()
+		tw = tar.NewWriter(gz)
+	} else {
+		tw = tar.NewWriter(destination)
+	}
+	defer func() {
+		// make sure to check the error on Close
+		if err := tw.Close(); err != nil {
+			log.Errorf("artifacts writer: error closing archive writer")
+			panic(err)
```

Comment:
I've been thinking about this as well. It might be tricky to do something else. 
Anyway, this is "just" writing part of library. So if you can not write artifact file you simply panic and then you will try once more. 
I am open for suggestions here.